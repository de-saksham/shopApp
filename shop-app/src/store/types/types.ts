export enum Category {
    VEGETABLES = 'VEGETABLES',
    FRUITS = 'FRUITS',
    CHEESE = 'CHEESE'
}
export interface CartProducts {
    _id: string;
    title: string;
    desc: string;
    price: number;
    stock: number;
    category: Category;
    quantity: number;
}

export interface Product {
    _id: string;
    title: string;
    desc: string;
    price: number;
    stock: number;
    category: Category
}

export interface State {
    isLoading: boolean;
    isSuccess: boolean;
    totalPages: number;
    currentPage: number;
    activeCategory: string;
    products: Product[];
    isCartEmpty: boolean;
    isModalActive: boolean;
    cart: CartProducts[];
    orderStatus: boolean;
}

export interface initApp {
    isSuccess: boolean,
    totalPages: number,
}

export interface storeCategory {
    category: string;
}

export interface storeProducts {
    products: Product[];
}

export interface addToCart {
    product: Product;
    quantity: number;
    variant: string;
}

export interface currentPage {
    currentPage: number;
}

export interface orderProducts {
    quantity: number;
    itemId: string;
}

export interface order {
    order: {
        userId: number,
        items: orderProducts[];
    }
}

export interface orderStatus {
    orderStatus: boolean;
}
