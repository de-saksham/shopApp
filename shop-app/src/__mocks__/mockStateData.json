{
    "reducer": {
        "ProductReducer": {
            "isLoading": false,
            "isSuccess": true,
            "totalPages": 5,
            "currentPage": 1,
            "activeCategory": "FRUITS",
            "isCartEmpty": false,
            "isModalActive": false,
            "orderStatus": true,
            "products": [
                {
                    "_id": "642899de1d0bfcc11aa6778a",
                    "title": "Washington Apple",
                    "desc": "Juicy red apples straight from United States of America",
                    "price": 7.99,
                    "stock": 2,
                    "category": "FRUITS"
                },
                {
                    "_id": "642899de1d0bfcc11aa6778b",
                    "title": "Green Apple",
                    "desc": "The Granny Smith, also known as a green apple or sour apple",
                    "price": 19.99,
                    "stock": 124,
                    "category": "FRUITS"
                }
            ],
            "cart": [
                {
                    "_id": "642899de1d0bfcc11aa6778a",
                    "title": "Washington Apple",
                    "desc": "Juicy red apples straight from United States of America",
                    "price": 7.99,
                    "stock": 2,
                    "category": "FRUITS",
                    "quantity": 1
                },
                {
                    "_id": "642899de1d0bfcc11aa6778b",
                    "title": "Green Apple",
                    "desc": "The Granny Smith, also known as a green apple or sour apple",
                    "price": 19.99,
                    "stock": 124,
                    "category": "FRUITS",
                    "quantity": 11
                }
            ]
        },
        "productApi": {
            "data": [
                {
                    "_id": "642899de1d0bfcc11aa6778a",
                    "title": "Washington Apple",
                    "desc": "Juicy red apples straight from United States of America",
                    "price": 7.99,
                    "stock": 2,
                    "category": "FRUITS"
                },
                {
                    "_id": "642899de1d0bfcc11aa6778b",
                    "title": "Green Apple",
                    "desc": "The Granny Smith, also known as a green apple or sour apple",
                    "price": 19.99,
                    "stock": 124,
                    "category": "FRUITS"
                }
            ]
        },
        "orderApi": {
            "data": {
                "order": [{
                    "itemId": "642899de1d0bfcc11aa6778b",
                    "quantity": 2
                }],
                "userId": 1
            }
        }
    }
}