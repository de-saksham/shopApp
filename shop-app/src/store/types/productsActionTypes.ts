import { TypeOfExpression } from 'typescript';
import * as actionTypes from '../actions/actionTypes';
import { CartProducts, Product } from './types';

export interface Init {
    type: typeof actionTypes.INIT;
    isSuccess: boolean;
    totalPages: number;
};

export interface IsLoading {
    type: typeof actionTypes.IS_LOADING;
}

export interface StoreCategory {
    type: typeof actionTypes.STORE_CATEGORY;
    category: string;
};

export interface StoreProducts {
    type: typeof actionTypes.STORE_PRODUCTS;
    products: Product[];
};

export interface AddToCart {
    type: typeof actionTypes.ADD_TO_CART;
    product: Product;
    quantity: number;
    variant: string;
};

export interface UpdateCurrentPage {
    type: typeof actionTypes.UPDATE_CURRENT_PAGE;
    currentPage: number;
}

export interface IsModalActive {
    type: typeof actionTypes.IS_MODAL_ACTIVE;
}

export interface OrderStatus {
    type: typeof actionTypes.ORDER_STATUS;
    orderStatus: boolean;
}

export type ProductsActionTypes = 
    | Init
    | IsLoading
    | StoreCategory
    | StoreProducts
    | AddToCart
    | UpdateCurrentPage
    | IsModalActive
    | OrderStatus;

export type ProductAction = ProductsActionTypes;
