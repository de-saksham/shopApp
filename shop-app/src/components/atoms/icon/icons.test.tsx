import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Icons from './icon';
import { Category } from '../../../store/types/types';
import configureStore from 'redux-mock-store';
import mockStateData from '../../../__mocks__/mockStateData.json';
import { Provider } from 'react-redux';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { configure } from 'enzyme';
import * as ShallowRenderer from 'react-test-renderer/shallow';

const mockState = mockStateData;
const mockStore = configureStore([]);
const store = mockStore(mockState);
configure({ adapter: new Adapter() });


describe('Icons - Test Suite', () => {
    let shallowRenderer: any;

    beforeEach(() => {
        shallowRenderer = ShallowRenderer.createRenderer();
    });

    test('Snapshot', () => {
        shallowRenderer.render(
            <Provider store={store}>
                <Icons iconName='back' locationType='header' text='' />
            </Provider>
        );
        expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
    });
});

describe('Icons - Header', () => {
  const mockNavigation = jest.fn();
  const mockCart = jest.fn();

  const defaultProps = {
    iconName: 'cart',
    locationType: 'header',
    text: 'Mock Text',
    activeCategory: 'VEGETABLES' as Category,
    navigation: mockNavigation,
    cart: mockCart,
  };

  beforeEach(() => {
    render(
      <Provider store={store}>
        <Icons  {...defaultProps}/>
      </Provider>
    );
  });

  it('renders icon', async() => {
    await waitFor(() => {
        expect(screen.getByAltText('cart')).toBeInTheDocument();
    });
  });

  it('renders header text', async() => {
    await waitFor(() => {
        expect(screen.getByText('Mock Text')).toBeInTheDocument();
    });
  });

  it('renders navigation text', async() => {
    await waitFor(() => {
        expect(screen.getByText('Mock Text')).toBeInTheDocument();
    });
  });

  it('renders active category text', async() => {
    await waitFor(() => {
        expect(screen.getByText('Mock Text')).toHaveClass('h4');
    });
  });

  it('calls navigation function on click', async() => {
    await waitFor(() => {
        fireEvent.click(screen.getByText('Mock Text'));
        expect(mockNavigation).toHaveBeenCalledWith('Mock Text');
    });
  });

  it('calls cart function on click', async() => {
    await waitFor(() => {
        fireEvent.click(screen.getByAltText('cart'));
        expect(mockCart).toHaveBeenCalled();
    });
  });
});

describe('Icons - Header', () => {
    const mockNavigation = jest.fn();
    const mockCart = jest.fn();

    const defaultProps = {
      iconName: 'cart',
      locationType: 'nav',
      text: 'Fruits',
      navigation: mockNavigation,
      cart: mockCart,
    };
  
    beforeEach(() => {
      render(
        <Provider store={store}>
          <Icons  {...defaultProps}/>
        </Provider>
      );
    });

    it('renders active category text', async() => {
        await waitFor(() => {
            expect(screen.getByText('Fruits')).toBeInTheDocument();
        });
      });

    
    it('should render the category as active if it is the activeCategory', async() => {
        await waitFor(() => {
            expect(screen.getByText('Fruits')).toHaveClass('activeCategory');
        });
    });  
  
    it('should render the category as inactive if it is not the activeCategory', async() => {
        const { getByText } = render(<Provider store={store}><Icons text={'Vegetables'} locationType='nav' iconName='cart' /></Provider>);
        await waitFor(() => {
            const category = getByText('Vegetables');
            expect(category).toHaveClass('inactiveCategory');
        });
    });
  });