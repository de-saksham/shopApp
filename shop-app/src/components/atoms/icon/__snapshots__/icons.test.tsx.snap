// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icons - Test Suite Snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "getServerState": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Icons
    iconName="back"
    locationType="header"
    text=""
  />
</Context.Provider>
`;
